src/BinSkim.Rules/PERules/BA6006.EnableLinkTimeCodeGeneration.cs

```diff
@@ -0,0 +1,93 @@
+    public class EnableLinkTimeCodeGeneration : WindowsBinaryAndPdbSkimmerBase
+    {
+        /// <summary>
+        /// BA6006
+        /// </summary>
+        public override string Id => RuleIds.EnableLinkTimeCodeGeneration;
+
+        /// <summary>
+        /// Enabling Link Time Code Generation (LTCG) performs whole-program optimization, which is able to better optimize code across
+        /// translation units. LTCG is also a prerequisite for Profile-Guided Optimization (PGO) which can further improve performance.
+        /// </summary>
+
+        public override MultiformatMessageString FullDescription =>
+            new MultiformatMessageString { Text = RuleResources.BA6006_EnableLinkTimeCodeGeneration_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA6006_Pass),
+                    nameof(RuleResources.BA6006_Warning),
+                    nameof(RuleResources.NotApplicable_InvalidMetadata)
+                };
+
+        private const string AnalyzerName = RuleIds.EnableLinkTimeCodeGeneration + "." + nameof(EnableLinkTimeCodeGeneration);
+
+        public override AnalysisApplicability CanAnalyzePE(PEBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            PE portableExecutable = target.PE;
+            AnalysisApplicability result = AnalysisApplicability.NotApplicableToSpecifiedTarget;
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsILOnlyAssembly;
+            if (portableExecutable.IsILOnly) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsResourceOnlyBinary;
+            if (portableExecutable.IsResourceOnly) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.CouldNotLoadPdb;
+            if (target.Pdb == null) { return result; }
+
+            Pdb pdb = target.Pdb;
+            reasonForNotAnalyzing = MetadataConditions.NotAReleaseBuild;
+            if (!portableExecutable.IsMostlyOptimized(pdb)) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsNotBuiltWithMSVC;
+            if (!portableExecutable.IsTargetCompiledWithMSVC(target.Pdb)) { return result; }
+
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        public override void AnalyzePortableExecutableAndPdb(BinaryAnalyzerContext context)
+        {
+            PEBinary target = context.PEBinary();
+            PE pe = target.PE;
+            Pdb pdb = target.Pdb;
+
+            if (!pe.IsLinkTimeCodeGenerationEnabled(pdb))
+            {
+                // '{0}' was compiled with without Link Time Code Generation (/LTCG). Enabling LTCG can improve optimizations and performance.
+                context.Logger.Log(this,
+                    RuleUtilities.BuildResult(ResultKind.Fail, context, null,
```

Comment:
>ResultKind.Fail [](http://example.com/codeflow?start=46&length=15)

Here's a subtlety, it's better to pass the failure level severity here rather than the result kind.

i.e., ResultKind.Fail can be one or error, warning, or informational. In this case, we fire a warning because that's the default.

You should prefer to pass FailureLevel.Warning here which makes the severity specific (this is a different logging overload provided for this purpose).

This note applies to your other Log statements in all rules.

Reply:
Ah I didn't realize there were different overloads.  Your suggestion more closely matches my intent.